plutus-metatheory/src/VerifiedCompilation/UCaseOfCase.lagda.md

```diff
@@ -0,0 +1,100 @@
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip ft ft') -- recursive translation for the false branch
+                ----------------------
+                → X ⊢̂
+                       (case ((((force (builtin ifThenElse)) · b) · (constr tn tt)) · (constr fn ft)) alts)
+                       ⊳̂ ((((force (builtin ifThenElse)) · b) · (case (constr tn' tt') alts')) · (case (constr tn' tt') alts'))
+   var : ∀ {x : X} → X ⊢̂ (` x) ⊳̂ (` x) 
+   ƛ   : ∀ {x x' : Maybe X ⊢}
+           → Maybe X ⊢̂ x ⊳̂ x'
+           ----------------------
+           →  X ⊢̂ ƛ x ⊳̂ ƛ x' 
+   _·_ : ∀ {f t f' t' : X ⊢} → (X ⊢̂ f ⊳̂ f') → (X ⊢̂ t ⊳̂ t') → (X ⊢̂ (f · t) ⊳̂ (f' · t'))
+   force : ∀ {t t' : X ⊢} → X ⊢̂ t ⊳̂ t' → X ⊢̂ force t ⊳̂ force t'  
+   delay : ∀ {t t' : X ⊢} → X ⊢̂ t ⊳̂ t' → X ⊢̂ delay t ⊳̂ delay t'  
+   con : ∀ {tc : TmCon} → X ⊢̂ con tc ⊳̂ con tc
+   constr : ∀ {xs xs' : List (X ⊢)} { n : ℕ }
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip xs xs')
+                ------------------------
+                → X ⊢̂ constr n xs ⊳̂ constr n xs' 
+   case :  ∀ {p p' : X ⊢} {alts alts' : List (X ⊢)}
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip alts alts') -- recursive translation for the other case patterns
+                → X ⊢̂ p ⊳̂ p'
+                ------------------------
+                → X ⊢̂ case p alts ⊳̂ case p alts' 
+   builtin : ∀ {b : Builtin} → X ⊢̂ builtin b ⊳̂ builtin b
+   error : X ⊢̂ error ⊳̂ error
+
+```
+
+## Decision Procedure
+
+Since this compiler phase is just a syntax re-arrangement in a very particular situation, we can
+match on that situation in the before and after ASTs and apply the translation rule for this, or
+expect anything else to be unaltered.
+
+This can just traverse the ASTs applying the constructors from the transition relation.
+
+```
+--_⊢̂_⊳̂?_ : {X : Set} (ast ast' : X ⊢) → Dec (X ⊢̂ ast ⊳̂ ast')
+--_⊢̂_⊳̂?_ ast ast' = ?
+
+```
+
+## Semantic Equivalence
+
+We can show that this translation never alters the semantics of the statement. This is shown
+in terms of the CEK machine evaluation. Since it is a simple re-arrangement of the syntax, it
+isn't a refinement argument - the state before and after the operation is the same type, and is
+unaltered buy the syntax re-arrangement.
+
+This does rely on the encoding of the semantics of `IfThenElse` in the CEK module, since we
+need to show that the effective list of cases is the same as it would have been without the re-arrangement.
+
+The `stepper` function uses the CEK machine to evaluate a term. Here we call it with a very
+large gas budget and begin in an empty context (which assumes the term is closed).
+
+```
+--semantic_equivalence : ∀ {X set} {ast ast' : ⊥ ⊢}
+ --                    → ⊥ ⊢ ast ⊳ ast'
+ -- <Some stuff about whether one runs out of gas -- as long as neither runs out of gas, _then_ they are equivilent> 
```

Comment:
Can we ignore gas problems? Why do we care about them? Some optimizations may slow the program down in edge cases while significantly improving performance on average and allowing other optimizations to fire.

Reply:
Jacco made the valid point that we will be doing inductive proofs so if the gas limit is finite we need to account for that or the induction won't work.

Reply:
Gas is just a specific way to look the non-termination effect (see [Turing-Completeness Totally Free](https://strathprints.strath.ac.uk/60166/1/McBride_LNCS2015_Turing_completeness_totally_free.pdf)), so it's not that you have to have it to do proofs, but I guess it's the simplest way to look at non-termination.

I'm a bit confused: my original comment was about "gas" in Plutus, i.e. CPU and MEM units, is this comment in the Agda code about a completely unrelated notion of gas that we need just for easy termination argument in Agda? In that case I'm totally fine with it.

Reply:
My problem is simpler: the CEK semantics has clauses for `gasError` so we have to handle those somehow in the induction proofs :)

Reply:
You could pretend it didn't exist if you wanted to, because the machine supports a budgeting mode that doesn't have any gas limitations.